src/AutoRest.CSharp/MgmtTest/Generation/ScenarioTestWriter.cs

```diff
@@ -0,0 +1,347 @@
+                        }
+                        writer.Line();
+                    }
+                }
+            }
+            writer.Line();
+        }
+
+        protected void WriteVariableInitializations(OavVariableScope ovs, IEnumerable<string> existedVarialbes)
+        {
+            IEnumerable<string> requiredVariables = ovs.RequiredVariables ?? new List<string>();
+            Dictionary<string, string> requiredVariableDefault = ovs.RequiredVariablesDefault ?? new Dictionary<string, string>();
+            var variables = ovs.Variables ?? new Dictionary<string, object?>();
+
+            if (ovs is TestDefinitionModel)
+            foreach (var requiredVariable in requiredVariables)
+            {
+                if (!existedVarialbes.Contains(requiredVariable))
+                {
+                    writer.Append($"string {new CodeWriterDeclaration(requiredVariable):D} = Environment.GetEnvironmentVariable(\"{requiredVariable.ToSnakeCase().ToUpper()}\")");
+                }
+                else
+                {
+                    writer.Append($"{requiredVariable} = Environment.GetEnvironmentVariable(\"{requiredVariable.ToSnakeCase().ToUpper()}\")");
+                }
+                if (requiredVariableDefault.ContainsKey(requiredVariable))
+                {
+                    writer.Append($" ?? {requiredVariableDefault[requiredVariable]:L}");
+                }
+                writer.Line($";");
+            }
+
+            foreach (var variable in variables.Keys)
+            {
+                var defaultValue = ovs.GetVariableDefaultValue(variable).RefScenariDefinedVariables(existedVarialbes);
+                if (!requiredVariables.Contains(variable) && !existedVarialbes.Contains(variable))
+                {
+                    writer.Line($"string {new CodeWriterDeclaration(variable):D} = {defaultValue};");
+                }
+                else
+                {
+                    writer.Line($"{variable} = {defaultValue};");
+                }
+            }
+
+            if (ovs is not TestDefinitionModel && variables.Count()>0)
+            {
+                writer.Append($"Console.WriteLine($\"{(ovs is Scenario? "Scenario": "Step")} variables: ");
+                foreach (var variable in variables.Keys)
+                {
+                    writer.Append($"{variable} -> $\\\"{{{variable}}}\\\", ");
+                }
+                writer.Line($"\");");
+            }
+            writer.Line();
+        }
+
+        protected void WriteTesterCtors()
+        {
+            // write protected default constructor
+            if (testDef.Scope == TestDefinitionModel.Scope_ResourceGroup)
+            {
+                writer.Line($"{typeof(ResourceGroup)} resourceGroup;");
+            }
+
+            if (!(testDef.RequiredVariables?.Contains("resourceGroupName") is true))
+            {
+                testDef.RequiredVariables?.Add("resourceGroupName");
```

Comment:
If you want to add something, you need to first ensure this is not null. Please do the null check before we are adding things